### ISSUE TITLE
phantomjs binary fails to install when using "npm install --ignore-scripts"

### ISSUE THREAD

**Issue by agenaille** [Medium/phantomjs#521]
On our project, we have recently been made aware of the [npm install vulnerability](http://news.softpedia.com/news/node-js-package-manager-vulnerable-to-malicious-worm-packages-502216.shtml) in which a malicious user could add some lifecycle scripts which execute postinstall with elevated permissions.

[Npm recommends a mitigation strategy](http://blog.npmjs.org/post/141702881055/package-install-scripts-vulnerability) of disabling scripts globally via:

`npm config set ignore-scripts true`

or

`npm install --ignore-scripts`

In our application, we are using phantomjs to execute karma unit tests, but when we download the dependencies with "npm install", and also using npm config set ignore-scripts true, the phantomjs binary fails to be recognized by the karma runner.

Now, I've found that I can workaround this by switching to the phantomjs directory and running the install script manually like so:

`//regularly install phantomjs with the other modules`
`npm install`

`//switch to phantomjs module dir`
`cd node_modules\phantomjs`

`//run the "install" script specified in package.json manually`
`npm run-script --ignore-scripts false install`

I'm wondering if this is a bug because one cannot install the module without needing to execute scripts which goes against npm guidance of disabling the scripts to avoid the worm issue?

If there is nothing that can be done, no worries, but at least someone else might find this info/workaround helpful.

Any help would be greatly appreciated! :) Thank you for your time, and also, excellent product!

**Comment by nicks** [comment]
yep. You can turn automatic installation scripts off, but then you'll have to run the installation script manually yourself. 

re: "npm guidance of disabling the scripts to avoid the worm issue": that is not quite accurate. From the official NPM blog post:

http://blog.npmjs.org/post/141702881055/package-install-scripts-vulnerability

> Installation and other lifecycle scripts are a useful tool that allows package authors to set up configuration, compile binary dependencies, and perform other actions that make using npm packages convenient.

> On balance, it’s npm’s belief that the utility of having installation scripts is greater than the risk of worms. This is a tradeoff that we will continue to evaluate.

**Comment by agenaille** [comment]
@nicks Thanks for your reply. I work for a certain clientele which forces our software to go through security penetration audits, and our clients enforce NIST and FIPS compliance and such, so these security vulnerabilities are treated with utmost concern. This phantomjs module has over 90 dependencies, and every single one of those dependencies has the possibility to specify a malicious script which will automatically execute after running "npm install." Also, we are using Express and tons of grunt modules, so realistically we have far more than 500 transitive dependencies that could specify a malicious script easily.

I agree that there is a pretty good track record of clean software on npm, and that all the great things that npm offers are super amazing (I love node and npm), but in our case, we have no choice but to mitigate the risk, as we don't want to be the next headline in the newspaper. 

What we've done is implemented a whitelist of packages and script names which must be executed so that we can do something like:

`npm install`
`node runPostInstallScriptsWhiteList`

This will ensure we only run the phantomjs install script, and no other install scripts from any of the many, many transitive dependencies, and will pretty much guarantee the security of the system and network.

Thanks!
